include/materials/tensor.tcc

```diff
@@ -0,0 +1,226 @@
+  Eigen::Matrix<double, 6, 1> dq_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
+  if (abs(deviatoric_q) > 1.E-6) {
+    dq_dsigma(0) = 3. / 2. / deviatoric_q * dev_stress(0);
+    dq_dsigma(1) = 3. / 2. / deviatoric_q * dev_stress(1);
+    dq_dsigma(2) = 3. / 2. / deviatoric_q * dev_stress(2);
+    dq_dsigma(3) = 3. / deviatoric_q * dev_stress(3);
+    dq_dsigma(4) = 3. / deviatoric_q * dev_stress(4);
+    dq_dsigma(5) = 3. / deviatoric_q * dev_stress(5);
+  }
+
+  return dq_dsigma;
+}
+
+//! Compute derivative of J2 in terms of stress sigma
+inline Eigen::Matrix<double, 6, 1> mpm::tensor::compute_dj2_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
+
+  // Compute dj2 / dsigma
+  Eigen::Matrix<double, 6, 1> dj2_dsigma = dev_stress;
+  dj2_dsigma(3) *= 2.0;
+  dj2_dsigma(4) *= 2.0;
+  dj2_dsigma(5) *= 2.0;
+
+  return dj2_dsigma;
+}
+
+//! Compute derivative of J3 in terms of stress sigma
+inline Eigen::Matrix<double, 6, 1> mpm::tensor::compute_dj3_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2
+  double j2 = compute_j2(stress);
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
+
+  // Compute dj3 / dsigma
+  Eigen::Matrix<double, 3, 1> dev1;
+  dev1(0) = dev_stress(0);
+  dev1(1) = dev_stress(3);
+  dev1(2) = dev_stress(5);
+  Eigen::Matrix<double, 3, 1> dev2;
+  dev2(0) = dev_stress(3);
+  dev2(1) = dev_stress(1);
+  dev2(2) = dev_stress(4);
+  Eigen::Matrix<double, 3, 1> dev3;
+  dev3(0) = dev_stress(5);
+  dev3(1) = dev_stress(4);
+  dev3(2) = dev_stress(2);
+
+  Eigen::Matrix<double, 6, 1> dj3_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
+  dj3_dsigma(0) = dev1.dot(dev1) - (2. / 3.) * j2;
+  dj3_dsigma(1) = dev2.dot(dev2) - (2. / 3.) * j2;
+  dj3_dsigma(2) = dev3.dot(dev3) - (2. / 3.) * j2;
+  dj3_dsigma(3) = 2.0 * dev1.dot(dev2);
+  dj3_dsigma(4) = 2.0 * dev2.dot(dev3);
+  dj3_dsigma(5) = 2.0 * dev1.dot(dev3);
+
+  return dj3_dsigma;
+}
+
+//! Compute derivative of Lode angle theta in terms of stress sigma
+inline Eigen::Matrix<double, 6, 1> mpm::tensor::compute_dtheta_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2
+  double j2 = compute_j2(stress);
+
+  // Compute J3
+  double j3 = compute_j3(stress);
+
+  // Compute dj2_dsigma
+  Eigen::Matrix<double, 6, 1> dj2_dsigma = compute_dj2_dsigma(stress);
+
+  // Compute dj3_dsigma
+  Eigen::Matrix<double, 6, 1> dj3_dsigma = compute_dj3_dsigma(stress);
+
+  // Declare R as zero to avoid division by zero J2
+  // R is defined as R = cos(3 theta)
+  double R = 0.0;
+
+  // Define derivatives of R in terms of J2 and J3
+  double dR_dj2 = -9.0 / 4.0 * sqrt(3.0) * j3;
+  double dR_dj3 = 3.0 / 2.0 * sqrt(3.0);
+
+  // Compute derivative of theta in terms of R
+  double dtheta_dR = -1.0 / 3.0;
+
+  // Update when J2 is non zero
+  if (abs(j2) > 1.0E-6) {
```

Comment:
Why 1.0E-6?

Reply:
We can change them of course, but I just chose this as a safe-guard. Note that since we are talking about Pa as unit for stress, it is rarely that we have stress so small that it would trigger this if condition. It usually only happens when we have zero stresses all throughout.

Should we change them to 1.E-16 instead?